pernosco-submit

```diff
@@ -629,76 +635,216 @@ end
+
+        Returns:
+            metadata (str): A string of comma-separated metadata values
+    """
+    # Send the public key with the signature so the server can easily
+    # determine which key was used and check that the key is authorized
+    metadata = "publickey=%s,signature=%s,user=%s,group=%s"%(strip_wrapper(public_key), signature, pernosco_user, pernosco_group)
+    if args.title != None:
+        metadata += ",title=%s"%urllib.parse.quote(args.title)
+    if args.url != None:
+        metadata += ",url=%s"%urllib.parse.quote(args.url)
+
+    if 'PERNOSCO_EXTRA_METADATA' in os.environ:
+        metadata += ",%s"%os.environ['PERNOSCO_EXTRA_METADATA']
+    return metadata
+
+def upload_file(nonce, signature, aws_access_key_id, aws_secret_access_key, public_key, pernosco_user, pernosco_group, metadata, payload_file_name):
+    """
+    Upload the specified payload file to Pernosco.
+
+        Parameters:
+            nonce (str): Nonce for uploaded file name
+            signature (str): Message digest for the package to be uploaded
+            aws_access_key_id (str): AWS access key ID
+            aws_secret_access_key (str): AWS secret access key
+            public_key (str): Public key for Pernosco user
+            pernosco_user (str): Pernosco user
+            pernosco_group (str): Pernosco group
+            metadata (str): Metadata required for the uplaod
+            payload_file_name (str): File to be uploaded
+    """
+    s3_url = "s3://pernosco-upload/%s.tar.zst"%nonce
+    really = ''
+    if args.subcommand == 'upload' and args.dry_run:
+        really = " (not really)"
+    print("Uploading %d bytes to %s...%s"%(os.path.getsize(payload_file_name), s3_url, really))
+
+    aws_env = get_aws_environment(aws_access_key_id, aws_secret_access_key)
+
+    checker_args = ["aws", "lambda", "invoke", "--function-name", "upload-credential-check", "--qualifier"]
+    if 'PERNOSCO_CREDENTIAL_CHECKER' in os.environ:
+        checker_args.extend([os.environ['PERNOSCO_CREDENTIAL_CHECKER']])
+    else:
+        checker_args.extend(["PROD"])
+    checker_args.extend(["--payload", "\"publickey=%s,user=%s,group=%s\""%(strip_wrapper(public_key), pernosco_user, pernosco_group), "/dev/null"])
+
+    aws_cmd = ["aws", "s3", "cp", "--metadata", metadata, payload_file_name, s3_url]
+
+    if args.subcommand == 'upload' and args.dry_run:
+        check_call(["cp", payload_file_name, args.dry_run])
+        cmd_obj = dict(checker_cmd=checker_args, aws_cmd=aws_cmd, aws_env=aws_env)
+        with open("%s.cmd"%args.dry_run, "w") as cmd:
+            print(json.dumps(cmd_obj), file=cmd)
+
+        # Create a file containing cryptographic data required to upload the
+        # package at a later time.
+        with open("%s.crypto"%args.dry_run, "w") as crypto:
+            print("%s\n"%(public_key.strip()), file=crypto)
+            print("%s\n"%(signature.strip()), file=crypto)
+            print("%s\n"%(nonce.strip()), file=crypto)
```

Comment:
I think it would be a little simpler and more maintainable if the `.crypto` file is JSON.

Reply:
Good idea.